windows_sync_objects/src/objectmanager.cpp

```diff
@@ -0,0 +1,397 @@
+/*
+ * Copyright (c) 2018 Trail of Bits, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "objectmanager.h"
+#include "winapi.h" // Definitions for undocumented Windows APIs
+
+#include "osquery/core/windows/wmi.h" // We need this for wstringToString
+#include <osquery/logger.h>
+
+#include <array>
+#include <cstdint>
+#include <sstream>
+#include <vector>
+
+namespace {
+struct DirectoryEntry final {
+  std::wstring name;
+  std::wstring type;
+};
+}
+
+namespace trailofbits {
+std::string GetErrorMessage(ULONG error_code) {
+  std::stringstream message;
+
+  char* buffer = nullptr;
+  if (FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM |
+                         FORMAT_MESSAGE_ALLOCATE_BUFFER |
+                         FORMAT_MESSAGE_IGNORE_INSERTS,
+                     nullptr,
+                     error_code,
+                     MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
+                     reinterpret_cast<LPSTR>(&buffer),
+                     0U,
+                     nullptr) == 0 ||
+      buffer == nullptr) {
+    message << std::hex << error_code;
+
+  } else {
+    message << buffer << " (0x" << std::hex << error_code << ")";
+    LocalFree(buffer);
+  }
+
+  return message.str();
+}
+
+void ListObDirectoryObjectsHelper(std::vector<DirectoryEntry>& entry_list,
+                                  HANDLE directory) {
+  entry_list.clear();
+
+  std::array<std::uint8_t, 4096> buffer;
+  ULONG entry_index = 0U;
+
+  while (true) {
+    ULONG response_size = 0U;
+    std::memset(buffer.data(), static_cast<int>(buffer.size()), 0);
+
+    auto status = NtQueryDirectoryObject(directory,
+                                         buffer.data(),
+                                         static_cast<ULONG>(buffer.size()),
+                                         FALSE,
+                                         entry_list.empty() ? TRUE : FALSE,
+                                         &entry_index,
+                                         &response_size);
+    if (NT_ERROR(status)) {
+      auto error_code = RtlNtStatusToDosError(status);
+      VLOG(1) << "Directory query failed with the following error: "
+              << GetErrorMessage(error_code);
+      break;
+    }
+
+    auto directory_entries =
+        reinterpret_cast<const OBJECT_DIRECTORY_INFORMATION*>(buffer.data());
+
+    for (std::size_t i = 0U; directory_entries[i].Name.Length != 0; i++) {

```

Comment:
Is it possible for one of the returned directory entry names to be an empty string? 

From `UNICODE_STRING` docs at MSDN:

>Windows 7, Windows Server 2008, Windows Vista, Windows Server 2003 and Windows XP:  When the Length structure member is zero and the MaximumLength structure member is 1, the Buffer structure member can be an empty string or contain solely a null character. This behavior changed beginning with Windows Server 2008 R2 and Windows 7 with SP1.

Reply:
I'm now checking all fields to make sure